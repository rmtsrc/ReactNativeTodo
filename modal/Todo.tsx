import React from 'react';

import { createContext, useState } from 'react';

type Task = { id: string; task: string; done: boolean };

type TodoList = Task[];

type TodoContext = {
  list: TodoList;
  add(task: string): void;
  remove(id: string): void;
  toggleDone(id: string): void;
};

export const TodoContext = createContext<Partial<TodoContext>>({});

const defaultTodoList = [
  {
    id: '1',
    task: 'Foo',
    done: true,
  },
  {
    id: '2',
    task: 'Bar',
    done: false,
  },
];

export function TodoProvider({ children }) {
  const [list, setList] = useState(defaultTodoList);

  return (
    <TodoContext.Provider
      value={{
        list,
        add: task =>
          setList([
            ...list,
            {
              id: `${list.length + 1}`,
              task,
              done: false,
            },
          ]),
        remove: id => setList(list.filter(item => item.id !== id)),
        toggleDone: id => {
          const index = list.findIndex(item => item.id === id);
          const item = list[index];
          setList([
            ...list.slice(0, index),
            { ...item, done: !item.done },
            ...list.slice(index + 1),
          ]);
        },
      }}
    >
      {children}
    </TodoContext.Provider>
  );
}
