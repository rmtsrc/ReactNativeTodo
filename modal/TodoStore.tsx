import React, { createContext, useReducer } from 'react';

type Task = { id: string; task: string; done: boolean };
type TodoList = Task[];
type TodoDispatch = (action: object) => void;

const defaultTodoList = [
  {
    id: '1',
    task: 'Foo',
    done: true,
  },
  {
    id: '2',
    task: 'Bar',
    done: false,
  },
];

const todoReducer = (list, action) => {
  switch (action.type) {
    case 'add':
      return [
        ...list,
        {
          id: `${list.length + 1}`,
          task: action.task,
          done: false,
        },
      ];

    case 'remove':
      return list.filter(item => item.id !== action.id);

    case 'toggleDone':
      const index = list.findIndex(item => item.id === action.id);
      const item = list[index];
      return [...list.slice(0, index), { ...item, done: !item.done }, ...list.slice(index + 1)];

    default:
      throw new Error(`Unknown action type: ${action.type}`);
  }
};

export const TodoStore = createContext<TodoList>([]);
export const TodoDispatch = createContext<TodoDispatch>(() => {});

export function TodoStoreProvider({ children }) {
  const [list, dispatch] = useReducer(todoReducer, defaultTodoList);

  return (
    <TodoDispatch.Provider value={dispatch}>
      <TodoStore.Provider value={list}>{children}</TodoStore.Provider>
    </TodoDispatch.Provider>
  );
}
