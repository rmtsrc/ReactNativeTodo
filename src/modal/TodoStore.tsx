import React, { createContext, useReducer, useEffect } from 'react';
import { AsyncStorage } from 'react-native';

type Task = { id: string; task: string; done: boolean };
type TodoList = Task[];
type TodoDispatch = (action: object) => void;

const defaultTodoList = [
  {
    id: '1',
    task: 'Example completed task',
    done: true,
  },
  {
    id: '2',
    task: 'Example todo',
    done: false,
  },
];

export const todoReducer = (list, action) => {
  let newList;
  switch (action.type) {
    case 'loaded': {
      newList = action.list === null ? [...defaultTodoList] : [...action.list];
      break;
    }
    case 'add': {
      newList = [
        ...list,
        {
          id: `${list.length + 1}`,
          task: action.task,
          done: false,
        },
      ];
      break;
    }
    case 'edit': {
      const index = list.findIndex(item => item.id === action.id);
      const item = list[index];
      newList = [...list.slice(0, index), { ...item, task: action.task }, ...list.slice(index + 1)];
      break;
    }
    case 'remove': {
      newList = list.filter(item => item.id !== action.id);
      break;
    }
    case 'toggleDone': {
      const index = list.findIndex(item => item.id === action.id);
      const item = list[index];
      newList = [...list.slice(0, index), { ...item, done: !item.done }, ...list.slice(index + 1)];
      break;
    }
    default: {
      throw new Error(`Unknown action type: ${action.type}`);
    }
  }

  async function persistData() {
    await AsyncStorage.setItem('todo-list', JSON.stringify(newList));
  }
  persistData();

  return newList;
};

export const TodoStore = createContext<TodoList>(null);
export const TodoDispatch = createContext<TodoDispatch>(() => {});

export function TodoStoreProvider({ children }) {
  const [list, dispatch] = useReducer(todoReducer, null);

  useEffect(() => {
    async function fetchData() {
      const list = await AsyncStorage.getItem('todo-list');
      dispatch({ type: 'loaded', list: JSON.parse(list) });
    }
    fetchData();
  }, []);

  return (
    <TodoDispatch.Provider value={dispatch}>
      <TodoStore.Provider value={list}>{children}</TodoStore.Provider>
    </TodoDispatch.Provider>
  );
}
