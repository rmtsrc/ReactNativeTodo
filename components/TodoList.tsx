import React from 'react';
import { FlatList, Text } from 'react-native';
import TodoListItem from './TodoListItem';
import { TodoContext } from '../modal/Todo';

export default function TodoList() {
  return (
    <TodoContext.Consumer>
      {context => (
        <FlatList
          data={context.list}
          keyExtractor={item => item.id}
          renderItem={({ item }) => (
            <TodoListItem
              id={item.id}
              task={item.task}
              checked={item.done}
              remove={context.remove}
              toggleDone={context.toggleDone}
            />
          )}
          ListEmptyComponent={
            <Text style={{ fontSize: 110, textAlign: 'center', marginTop: 40 }}>
              ðŸ“‹âœ…
            </Text>
          }
        />
      )}
    </TodoContext.Consumer>
  );
}
