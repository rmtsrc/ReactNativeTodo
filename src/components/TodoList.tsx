import React from 'react';
import { FlatList, Text } from 'react-native';
import TodoListItem from './TodoListItem';
import { TodoStore } from '../modal/TodoStore';

export default function TodoList() {
  return (
    <TodoStore.Consumer>
      {list =>
        list !== null && (
          <FlatList
            data={list}
            keyExtractor={item => item.id}
            renderItem={({ item }) => <TodoListItem id={item.id} task={item.task} checked={item.done} />}
            ListEmptyComponent={<Text style={{ fontSize: 110, textAlign: 'center', marginTop: 40 }}>ðŸ“‹âœ…</Text>}
          />
        )
      }
    </TodoStore.Consumer>
  );
}
