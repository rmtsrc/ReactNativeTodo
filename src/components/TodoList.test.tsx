import React from 'react';
import renderer from 'react-test-renderer';

import TodoList from './TodoList';
import { TodoStore } from '../modal/TodoStore';

describe('<TodoList />', () => {
  test('renders correctly when no data is given', () => {
    let wrapper = renderer.create(<TodoList />);
    expect(wrapper.toJSON()).toBeNull();
  });

  test('renders correctly when an empty todo list is given', () => {
    let wrapper = renderer.create(
      <TodoStore.Provider value={[]}>
        <TodoList />
      </TodoStore.Provider>
    );

    const {children} = wrapper.root.findByType('Text').props;
    expect(children).toBe('ðŸ“‹âœ…');

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('renders correctly when a valid todo list is given', () => {
    let wrapper = renderer.create(
      <TodoStore.Provider value={[{ id: '1', task: 'my task', done: false }]}>
        <TodoList />
      </TodoStore.Provider>
    );

    const { children } = wrapper.root.findByProps({ testID: 'listItemTitle' }).props;
    expect(children).toBe('my task');
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
