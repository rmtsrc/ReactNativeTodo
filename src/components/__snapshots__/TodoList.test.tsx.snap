// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoList /> renders correctly when a valid todo list is given 1`] = `
<RCTScrollView
  ListEmptyComponent={
    <Text
      style={
        Object {
          "fontSize": 110,
          "marginTop": 40,
          "textAlign": "center",
        }
      }
    >
      ðŸ“‹âœ…
    </Text>
  }
  data={
    Array [
      Object {
        "done": false,
        "id": "1",
        "task": "my task",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "borderBottomWidth": 0.5,
              "borderColor": "#bcbbc1",
              "flexDirection": "row",
              "padding": 14,
            }
          }
        >
          <View
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "margin": 5,
                "marginLeft": 10,
                "marginRight": 10,
                "opacity": 1,
                "padding": 10,
              }
            }
            testID="checkbox"
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                }
              }
            >
              <Text />
            </View>
          </View>
          <View
            width={0}
          />
          <View
            style={
              Object {
                "flex": 1,
                "justifyContent": "center",
              }
            }
          >
            <Text
              style={
                Object {
                  "backgroundColor": "transparent",
                  "fontSize": 17,
                }
              }
              testID="listItemTitle"
              theme={
                Object {
                  "colors": Object {
                    "disabled": "hsl(208, 8%, 90%)",
                    "divider": "#bcbbc1",
                    "error": "#ff190c",
                    "grey0": "#393e42",
                    "grey1": "#43484d",
                    "grey2": "#5e6977",
                    "grey3": "#86939e",
                    "grey4": "#bdc6cf",
                    "grey5": "#e1e8ee",
                    "greyOutline": "#bbb",
                    "platform": Object {
                      "android": Object {
                        "error": "#f44336",
                        "primary": "#2196f3",
                        "secondary": "#9C27B0",
                        "success": "#4caf50",
                        "warning": "#ffeb3b",
                      },
                      "ios": Object {
                        "error": "#ff3b30",
                        "primary": "#007aff",
                        "secondary": "#5856d6",
                        "success": "#4cd964",
                        "warning": "#ffcc00",
                      },
                    },
                    "primary": "#2089dc",
                    "searchBg": "#303337",
                    "secondary": "#8F0CE8",
                    "success": "#52c41a",
                    "warning": "#faad14",
                  },
                }
              }
            >
              my task
            </Text>
          </View>
          <View
            width={0}
          />
          <View>
            <View
              accessible={true}
              focusable={true}
              isTVSelectable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "justifyContent": "center",
                }
              }
            >
              <Text />
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`<TodoList /> renders correctly when an empty todo list is given 1`] = `
<RCTScrollView
  ListEmptyComponent={
    <Text
      style={
        Object {
          "fontSize": 110,
          "marginTop": 40,
          "textAlign": "center",
        }
      }
    >
      ðŸ“‹âœ…
    </Text>
  }
  data={Array []}
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <Text
      onLayout={[Function]}
      style={
        Object {
          "fontSize": 110,
          "marginTop": 40,
          "textAlign": "center",
        }
      }
    >
      ðŸ“‹âœ…
    </Text>
  </View>
</RCTScrollView>
`;
